README.md
#Behance block

##CONTENTS OF THIS FILE

  * Introduction
  * Requirements
  * Installation
  * Configuration
  * Using the module
  * Author

##INTRODUCTION

The Behance Block Drupal 8 module allows you to pull content from Behance and
display projects on your own website. Behance is the leading online platform to
showcase creative work, and by using this module you will get access to the data
behind millions of projects. Since Behance limits API calls, this module stores
retrieved data in the 'cache' file. The cache file is updated every 24 hours.

##REQUIREMENTS

None.

##INSTALLATION

1. Install the module as usual.
2. Go to "Extend" and enable the Behance Block module.

##CONFIGURATION

In order to get access to the Behance Block you need a unique API key. Go to
(https://www.behance.net/dev/register), and register your app. After that, go to
"Configuration" -> "WEB SERVICES" -> "Behance Block" and enter your API key and
portfolio owner's Username or ID.

##USING THE MODULE

1. Create a basic page.
2. Add a Behance Block to this page by going to "Structure" -> "Block layout"
and then click on the Place block button.

###AUTHOR

Goran Nikolovski
Website: (http://www.gorannikolovski.com)
Drupal: (https://www.drupal.org/user/3451979)
Email: nikolovski84@gmail.com

Company: Studio Present, Subotica, Serbia
Website: (http://www.studiopresent.com)
Drupal: (https://www.drupal.org/studio-present)
Email: info@studiopresent.com


CHANGELOG.txt
Behance Block module 8.x-1.x, 2016-10-06
----------------------------------------
- Initial release


src/Form/BehanceSettingsForm.php
<?php

namespace Drupal\behance_block\Form;

use Drupal\Core\Form\ConfigFormBase;
use Drupal\Core\Form\FormStateInterface;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;

/**
 * @file
 * Contains \Drupal\behance_block\Form\BehanceSettingsForm.
 */

/**
 * Defines a form that configures behance block settings.
 */
class BehanceSettingsForm extends ConfigFormBase {

  /**
   * {@inheritdoc}
   */
  public function getFormId() {
    return 'behance_block_settings';
  }

  /**
   * {@inheritdoc}
   */
  protected function getEditableConfigNames() {
    return array('behance_block.settings');
  }

  /**
   * {@inheritdoc}
   */
  public function buildForm(array $form, FormStateInterface $form_state) {

    $config = $this->config('behance_block.settings');

    // API key field.
    $form['api_key'] = array(
      '#type' => 'textfield',
      '#title' => t('API Key'),
      '#required' => TRUE,
      '#description' => t('Enter your API key. If you don\'t have one, visit <a target="_blank" href="https://www.behance.net/dev/register">this page</a> and get your Behance API key.'),
      '#default_value' => $config->get('api_key'),
    );

    // User ID or username field.
    $form['user_id'] = array(
      '#type' => 'textfield',
      '#title' => t('Username or User ID'),
      '#required' => TRUE,
      '#description' => t("Enter portfolio owner's ID or username."),
      '#default_value' => $config->get('user_id'),
    );

    // New tab checkbox.
    $form['new_tab'] = array(
      '#type' => 'checkbox',
      '#title' => t('Open links in new tab'),
      '#description' => t('Check if you want Behance links to be opened in the new tab.'),
      '#default_value' => $config->get('new_tab'),
    );

    return parent::buildForm($form, $form_state);

  }

  /**
   * {@inheritdoc}
   */
  public function validateForm(array &$form, FormStateInterface $form_state) {

    $api_key = $form_state->getValue('api_key');
    $user_id = $form_state->getValue('user_id');

    $isDataValid = $this->isDataValid($api_key, $user_id);

    if ($isDataValid == 403) {
      $form_state->setErrorByName('api_key', $this->t('Your API key is not valid'));
    }

    if ($isDataValid == 404) {
      $form_state->setErrorByName('user_id', $this->t('User not found'));
    }

  }

  /**
   * {@inheritdoc}
   */
  public function submitForm(array &$form, FormStateInterface $form_state) {

    $this->clearCacheFiles();

    $values = $form_state->getValues();

    $config = $this->config('behance_block.settings');
    $config->set('api_key', $values['api_key']);
    $config->set('user_id', $values['user_id']);
    $config->set('new_tab', $values['new_tab']);
    $config->save();

    parent::submitForm($form, $form_state);

  }

  /**
   * Check if API key and User ID are valid.
   */
  private function isDataValid($api_key, $user_id) {

    $client = new Client();

    try {
      $response = $client->get('https://api.behance.net/v2/users/' . $user_id . '?client_id=' . $api_key);
      $response_code = $response->getStatusCode();
    }
    catch (ClientException $e) {
      $response = $e->getResponse();
      $response_code = $response->getStatusCode();
      watchdog_exception('behance_block', $e);
    }

    // Valid (200 = OK).
    if ($response_code == 200) {
      return 200;
    }
    // API key in not valid.
    elseif ($response_code == 403) {
      return 403;
    }
    // User not found.
    elseif ($response_code == 404) {
      return 404;
    }

  }

  /**
   * Delete cache files.
   */
  private function clearCacheFiles() {

    if (file_exists('public://behance_fields.json')) {
      unlink('public://behance_fields.json');
    }

    if (file_exists('public://behance_projects.json')) {
      unlink('public://behance_projects.json');
    }

  }

}


src/Plugin/Block/BehanceBlock.php
<?php

namespace Drupal\behance_block\Plugin\Block;

use Drupal\Core\Block\BlockBase;
use Symfony\Component\DependencyInjection\ContainerInterface;
use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
use Drupal\Core\Config\ConfigFactoryInterface;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\ClientException;

/**
 * @file
 * Contains \Drupal\behance_block\Plugin\Block\BehanceBlock.
 */

/**
 * Provides a 'Behance Block' Block.
 *
 * @Block(
 *   id = "behance_block",
 *   admin_label = @Translation("Behance Block"),
 *   category = @Translation("Integrations"),
 * )
 */
class BehanceBlock extends BlockBase implements ContainerFactoryPluginInterface {

  /**
   * The config_factory.
   *
   * @var \Drupal\Core\Config\config_factoryInterface
   */
  protected $configFactory;

  private $apiKey;
  private $userId;
  private $newTab;
  private $behanceFieldsDate;
  private $behanceProjectsDate;

  /**
   * Class Constructor.
   */
  public function __construct(array $configuration, $plugin_id, $plugin_definition, ConfigFactoryInterface $configFactory) {
    parent::__construct($configuration, $plugin_id, $plugin_definition);
    $this->config_factory = $configFactory;
  }

  /**
   * {@inheritdoc}
   */
  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
    return new static(
      $configuration, $plugin_id, $plugin_definition, $container->get('config.factory')
    );
  }

  /**
   * {@inheritdoc}
   */
  public function build() {

    $this->api_key = $this->config_factory->get('behance_block.settings')->get('api_key');
    $this->user_id = $this->config_factory->get('behance_block.settings')->get('user_id');
    $this->new_tab = $this->config_factory->get('behance_block.settings')->get('new_tab');
    $this->behance_fields_date = $this->config_factory->get('behance_block.settings')->get('behance_fields_date');
    $this->behance_projects_date = $this->config_factory->get('behance_block.settings')->get('behance_projects_date');

    $output = array();

    $is_api_key_set = (isset($this->api_key) && !empty($this->api_key));
    $is_user_id_set = (isset($this->user_id) && !empty($this->user_id));

    // API key and User ID are set - show Behance projects.
    if ($is_api_key_set && $is_user_id_set) {

      $output[] = array(
        '#theme' => 'behance_block',
        '#projects' => $this->content(),
        '#tags' => $this->tags(),
        '#new_tab' => $this->getNewTab(),
        '#cache' => array('max-age' => 0),
        '#attached' => array('library' => array('behance_block/behance_block')),
      );

    }
    // Show error if required values are missing.
    else {

      $output[] = array(
        '#markup' => 'You must set an API key and username in the module settings. <a href="/admin/config/services/behance">Click here</a> to go the module settings.',
        '#cache' => array('max-age' => 0),
        '#attached' => array('library' => array('behance_block/behance_block')),
      );

    }

    return $output;

  }

  /**
   * Returns array with all projects.
   */
  private function content() {

    $projects_json = '';

    // Get projects from JSON cache file.
    $behance_projects_file_exists = file_exists('public://behance_projects.json');
    $behance_projects_file_date = $this->behance_projects_date;

    if (!$behance_projects_file_exists || $behance_projects_file_date != date('d.m.Y')) {
      $this->downloadProjectsJson();
    }

    if (file_exists('public://behance_projects.json')) {

      $json = file_get_contents('public://behance_projects.json');
      $projects_json = json_decode($json, TRUE);

    }

    return $projects_json;

  }

  /**
   * Open links in new tab or not.
   */
  private function getNewTab() {

    if ($this->newTab == 0) {
      return 'target=_self';
    }
    else {
      return 'target=_blank';
    }

  }

  /**
   * Returns all Behance tags in array.
   */
  private function tags() {

    $tags_json_array = array();

    // Get Behance field names from JSON cache file.
    $behance_fields_file_exists = file_exists('public://behance_fields.json');
    $behance_fields_file_date = $this->behance_fields_date;

    if (!$behance_fields_file_exists || $behance_fields_file_date != date('d.m.Y')) {
      $this->downloadFieldsJson();
    }

    if (file_exists('public://behance_fields.json')) {

      $tags = file_get_contents('public://behance_fields.json');
      $tags_json = json_decode($tags, TRUE);

      foreach ($tags_json['fields'] as $value) {
        $tags_json_array[$value['name']] = $value['id'];
      }

    }

    return $tags_json_array;

  }

  /**
   * Get Behance field names (tags) and store them in JSON file.
   */
  private function downloadFieldsJson() {

    // Get response from endpoint and save it.
    $client = new Client();

    try {
      $response = $client->get('http://api.behance.net/v2/fields?api_key=' . $this->api_key);
      $response_code = $response->getStatusCode();
      $behance_fields_json = $response->getBody();
    }
    catch (ClientException $e) {
      $response = $e->getResponse();
      $response_code = $response->getStatusCode();
      watchdog_exception('behance_block', $e);
    }

    if ($response_code == 200) {

      file_put_contents('public://behance_fields.json', $behance_fields_json);

      // Save date when the file is downloaded.
      $this->config_factory->getEditable('behance_block.settings')->set('behance_fields_date', date('d.m.Y'))->save();

    }

  }

  /**
   * Get Behance projects and store them in JSON file.
   */
  private function downloadProjectsJson() {

    // Get response from endpoint and save it.
    $projects_json = $this->getProjectsJson();

    if ($projects_json) {

      file_put_contents('public://behance_projects.json', $projects_json);

      // Save date when the file is downloaded.
      $this->config_factory->getEditable('behance_block.settings')->set('behance_projects_date', date('d.m.Y'))->save();

    }

  }

  /**
   * Get Behance projects from API endpoint.
   */
  private function getProjectsJson() {

    $i = 1;
    $loop_through = TRUE;
    $projects_json_full = array();

    $client = new Client();

    // Loop while you get not empty JSON response.
    while ($loop_through) {

      try {
        $response = $client->get('http://api.behance.net/v2/users/' . $this->user_id . '/projects?api_key=' . $this->api_key . '&per_page=24&page=' . $i);
        $response_code = $response->getStatusCode();
        $projects_json_page = $response->getBody();
        $projects_json = json_decode($projects_json_page, TRUE);
      }
      catch (ClientException $e) {
        $response = $e->getResponse();
        $response_code = $response->getStatusCode();
        watchdog_exception('behance_block', $e);
      }

      if ($response_code == 200) {
        if ($projects_json['projects']) {
          foreach ($projects_json['projects'] as $projects) {
            $projects_json_full[] = $projects;
          }
          $i++;
        }
        else {
          $loop_through = FALSE;
        }
      }
      else {
        $loop_through = FALSE;
      }

    }

    if ($projects_json_full) {
      return json_encode($projects_json_full);
    }
    else {
      return FALSE;
    }

  }

}
